spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessor.java

```diff
@@ -334,6 +365,66 @@ private String resolve(String value) {
 		return value;
 	}
 
+	private void registerBeansForDeclaration(RabbitListener rabbitListener) {
+		if (this.beanFactory instanceof ConfigurableBeanFactory) {
+			for (QueueBinding binding : rabbitListener.bindings()) {
+				org.springframework.amqp.rabbit.annotation.Queue bindingQueue = binding.value();
+				String queueName = (String) resolveExpression(bindingQueue.value());
+				Queue queue = new Queue(queueName,
+						resolveExpressionAsBoolean(bindingQueue.durable()),
+						resolveExpressionAsBoolean(bindingQueue.exclusive()),
+						resolveExpressionAsBoolean(bindingQueue.autoDelete()));
+				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(queueName + ++this.increment, queue);
+				Exchange exchange = null;
+				org.springframework.amqp.rabbit.annotation.Exchange bindingExchange = binding.exchange();
+				String exchangeName = (String) resolveExpression(bindingExchange.value());
+				String exchangeType = bindingExchange.type();
+				Binding actualBinding = null;
+				Object key = resolveExpression(binding.key());
+				if (!(key instanceof String)) {
+					throw new BeanInitializationException("key must resolved to a String, not: " + key.getClass().toString());
+				}
+ 				String resolvedKey = (String) key;
+				if (exchangeType.equals(ExchangeTypes.DIRECT)) {
+					exchange = new DirectExchange(exchangeName,
+							resolveExpressionAsBoolean(bindingExchange.durable()),
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey, null);
+				}
+				else if (exchangeType.equals(ExchangeTypes.FANOUT)) {
+					exchange = new FanoutExchange(exchangeName,
+							resolveExpressionAsBoolean(bindingExchange.durable()),
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, null, null);
+				}
+				else if (exchangeType.equals(ExchangeTypes.TOPIC)) {
+					exchange = new TopicExchange(exchangeName,
+							resolveExpressionAsBoolean(bindingExchange.durable()),
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey, null);
+				}
+				else {
+					throw new BeanInitializationException("Unexpected exchange type: " + exchangeType);
+				}
+				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment, exchange);
+				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment, actualBinding);
```

Comment:
I didn't want to inadvertently overwrite some other bean definition, perhaps the bean name should be even more distinct.